### Tranning Process of Bike Availability Model and Bike Stands Availability Model

## Setup and Initial Data Checks

In [5]:
import pandas as pd

In [6]:

df = pd.read_csv('input_data/final_merged_data.csv')
df.head(10)

,last_reported,station_id,num_bikes_available,num_docks_available,is_installed,is_renting,is_returning,name,address,lat,...,min_humidity_quality_indicator,min_relative_humidity_percent,humidity_std_quality_indicator,relative_humidity_std_deviation,max_pressure_quality_indicator,max_barometric_pressure_hpa,min_pressure_quality_indicator,min_barometric_pressure_hpa,pressure_std_quality_indicator,barometric_pressure_std_deviation
0,2024-12-01 00:10:00,10,15,1,True,True,True,DAME STREET,Dame Street,53.344006,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
1,2024-12-01 00:10:00,100,17,8,True,True,True,HEUSTON BRIDGE (SOUTH),Heuston Bridge (South),53.347107,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
2,2024-12-01 00:10:00,109,20,9,True,True,True,BUCKINGHAM STREET LOWER,Buckingham Street Lower,53.353333,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
3,2024-12-01 00:10:00,11,1,29,True,True,True,EARLSFORT TERRACE,Earlsfort Terrace,53.334293,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
4,2024-12-01 00:10:00,114,4,36,True,True,True,WILTON TERRACE (PARK),Wilton Terrace (Park),53.333652,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
5,2024-12-01 00:10:00,116,2,28,True,True,True,BROADSTONE,Broadstone,53.354700,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
6,2024-12-01 00:10:00,13,0,30,True,True,True,FITZWILLIAM SQUARE WEST,Fitzwilliam Square West,53.336075,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
7,2024-12-01 00:10:00,14,21,9,True,True,True,FOWNES STREET UPPER,Fownes Street Upper,53.344604,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
8,2024-12-01 00:10:00,15,1,15,True,True,True,HARDWICKE STREET,Hardwicke Street,53.355473,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
9,2024-12-01 00:10:00,17,3,17,True,True,True,GOLDEN LANE,Golden Lane,53.340800,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083


Take a glance of raw training data.

In [7]:
df = pd.read_csv('input_data/local_databasejcdecaux_availability.csv')
df.head(5)

,number,available_bike_stands,available_bikes,status,last_update
0,42,25,5,OPEN,1771077067000
1,30,12,8,OPEN,1771077223000
2,54,20,13,OPEN,1771076933000
3,108,21,14,OPEN,1771077214000
4,20,29,1,OPEN,1771077216000


In [8]:
df = pd.read_csv('input_data/local_databasejcdecaux_station.csv')
df.head(5)

,number,contract_name,name,address,bike_stands,lat,lng,banking,bonus
0,1,dublin,CLARENDON ROW,Clarendon Row,31,53.3409,-6.26250,0,0
1,2,dublin,BLESSINGTON STREET,Blessington Street,20,53.3568,-6.26814,0,0
2,3,dublin,BOLTON STREET,Bolton Street,20,53.3512,-6.26986,0,0
3,4,dublin,GREEK STREET,Greek Street,20,53.3469,-6.27298,0,0
4,5,dublin,CHARLEMONT PLACE,Charlemont Street,40,53.3307,-6.26000,0,0


In [9]:
df = pd.read_csv('input_data/local_databasejcdecaux_current.csv')
df.head(5)

,dt,feels_like,humidity,pressure,sunrise,sunset,temp,uvi,weather_id,wind_gust,wind_speed,rain_1h,snow_1h
0,2026-02-11 18:16:32,282.58,94,974,2026-02-11 07:51:51,2026-02-11 17:26:51,282.58,NaN,803,NaN,1.03,NaN,NaN
1,2026-02-12 16:24:11,277.05,97,986,2026-02-12 07:49:54,2026-02-12 17:28:53,279.86,NaN,501,NaN,4.12,3.16,NaN
2,2026-02-12 17:55:25,276.41,97,986,2026-02-12 07:49:54,2026-02-12 17:28:53,279.81,NaN,803,NaN,5.36,NaN,NaN
3,2026-02-12 18:55:11,276.09,97,986,2026-02-12 07:49:54,2026-02-12 17:28:53,279.81,NaN,803,NaN,6.17,NaN,NaN
4,2026-02-12 19:51:40,275.89,95,986,2026-02-12 07:49:54,2026-02-12 17:28:53,279.81,NaN,501,NaN,6.71,0.86,NaN


Take a glacen of the 3 tables that acquired, using this to deciding the method to manipulate the raw data (accoring to the data feature availability). (Our own 3 table cvs is not the full one, only export from the mysql to see the column, thus only 5 lines are exported.)

In [10]:
df = pd.read_csv('input_data/final_merged_data.csv')

with pd.option_context('display.max_columns', None):
    display(df.head())


print(f"Successfully loaded initial dataset with {df.shape[0]} rows and {df.shape[1]} columns")
print(f"Number of rows (instances): {df.shape[0]}")
print(f"Number of columns (features): {df.shape[1]}")

display(df.head())

,last_reported,station_id,num_bikes_available,num_docks_available,is_installed,is_renting,is_returning,name,address,lat,lon,capacity,stno,year,month,day,hour,minute,max_air_temp_quality_indicator,max_air_temperature_celsius,min_air_temp_quality_indicator,min_air_temperature_celsius,air_temp_std_quality_indicator,air_temperature_std_deviation,max_grass_temp_quality_indicator,max_grass_temperature_celsius,min_grass_temp_quality_indicator,min_grass_temperature_celsius,grass_temp_std_quality_indicator,grass_temperature_std_deviation,max_soil_temp_5cm_quality_indicator,max_soil_temperature_5cm_celsius,min_soil_temp_5cm_quality_indicator,min_soil_temperature_5cm_celsius,soil_temp_std_5cm_quality_indicator,soil_temperature_std_deviation_5cm,max_soil_temp_10cm_quality_indicator,max_soil_temperature_10cm_celsius,min_soil_temp_10cm_quality_indicator,min_soil_temperature_10cm_celsius,soil_temp_std_10cm_quality_indicator,soil_temperature_std_deviation_10cm,max_soil_temp_20cm_quality_indicator,max_soil_temperature_20cm_celsius,min_soil_temp_20cm_quality_indicator,min_soil_temperature_20cm_celsius,soil_temp_std_20cm_quality_indicator,soil_temperature_std_deviation_20cm,max_earth_temp_30cm_quality_indicator,max_earth_temperature_30cm_celsius,min_earth_temp_30cm_quality_indicator,min_earth_temperature_30cm_celsius,earth_temp_std_30cm_quality_indicator,earth_temperature_std_deviation_30cm,max_earth_temp_50cm_quality_indicator,max_earth_temperature_50cm_celsius,min_earth_temp_50cm_quality_indicator,min_earth_temperature_50cm_celsius,earth_temp_std_50cm_quality_indicator,earth_temperature_std_deviation_50cm,max_earth_temp_100cm_quality_indicator,max_earth_temperature_100cm_celsius,min_earth_temp_100cm_quality_indicator,min_earth_temperature_100cm_celsius,earth_temp_std_100cm_quality_indicator,earth_temperature_std_deviation_100cm,max_humidity_quality_indicator,max_relative_humidity_percent,min_humidity_quality_indicator,min_relative_humidity_percent,humidity_std_quality_indicator,relative_humidity_std_deviation,max_pressure_quality_indicator,max_barometric_pressure_hpa,min_pressure_quality_indicator,min_barometric_pressure_hpa,pressure_std_quality_indicator,barometric_pressure_std_deviation
0,2024-12-01 00:10:00,10,15,1,True,True,True,DAME STREET,Dame Street,53.344006,-6.266802,16,175,2024,12,1,0,10,0,14.01,0,13.9,0,0.033,0,11.93,0,11.73,0,0.056,0,10.15,0,10.11,0,0.008,0,9.58,0,9.56,0,0.005,0,8.83,0,8.8,0,0.007,0,8.88,0,8.85,0,0.008,0,8.5,0,8.47,0,0.008,0,9.87,0,9.83,0,0.008,0,84.3,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
1,2024-12-01 00:10:00,100,17,8,True,True,True,HEUSTON BRIDGE (SOUTH),Heuston Bridge (South),53.347107,-6.292041,25,175,2024,12,1,0,10,0,14.01,0,13.9,0,0.033,0,11.93,0,11.73,0,0.056,0,10.15,0,10.11,0,0.008,0,9.58,0,9.56,0,0.005,0,8.83,0,8.8,0,0.007,0,8.88,0,8.85,0,0.008,0,8.5,0,8.47,0,0.008,0,9.87,0,9.83,0,0.008,0,84.3,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
2,2024-12-01 00:10:00,109,20,9,True,True,True,BUCKINGHAM STREET LOWER,Buckingham Street Lower,53.353333,-6.249319,29,175,2024,12,1,0,10,0,14.01,0,13.9,0,0.033,0,11.93,0,11.73,0,0.056,0,10.15,0,10.11,0,0.008,0,9.58,0,9.56,0,0.005,0,8.83,0,8.8,0,0.007,0,8.88,0,8.85,0,0.008,0,8.5,0,8.47,0,0.008,0,9.87,0,9.83,0,0.008,0,84.3,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
3,2024-12-01 00:10:00,11,1,29,True,True,True,EARLSFORT TERRACE,Earlsfort Terrace,53.334293,-6.258503,30,175,2024,12,1,0,10,0,14.01,0,13.9,0,0.033,0,11.93,0,11.73,0,0.056,0,10.15,0,10.11,0,0.008,0,9.58,0,9.56,0,0.005,0,8.83,0,8.8,0,0.007,0,8.88,0,8.85,0,0.008,0,8.5,0,8.47,0,0.008,0,9.87,0,9.83,0,0.008,0,84.3,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
4,2024-12-01 00:10:00,114,4,36,True,True,True,WILTON TERRACE (PARK),Wilton Terrace (Park),53.333652,-6.248345,40,175,2024,12,1,0,10,0,14.01,0,13.9,0,0.033,0,11.93,0,11.73,0,0.056,0,10.15,0,10.11,0,0.008,0,9.58,0,9.56,0,0.005,0,8.83,0,8.8,0,0.007,0,8.88,0,8.85,0,0.008,0,8.5,0,8.47,0,0.008,0,9.87,0,9.83,0,0.008,0,84.3,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083


Successfully loaded initial dataset with 298946 rows and 78 columns
Number of rows (instances): 298946
Number of columns (features): 78


,last_reported,station_id,num_bikes_available,num_docks_available,is_installed,is_renting,is_returning,name,address,lat,...,min_humidity_quality_indicator,min_relative_humidity_percent,humidity_std_quality_indicator,relative_humidity_std_deviation,max_pressure_quality_indicator,max_barometric_pressure_hpa,min_pressure_quality_indicator,min_barometric_pressure_hpa,pressure_std_quality_indicator,barometric_pressure_std_deviation
0,2024-12-01 00:10:00,10,15,1,True,True,True,DAME STREET,Dame Street,53.344006,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
1,2024-12-01 00:10:00,100,17,8,True,True,True,HEUSTON BRIDGE (SOUTH),Heuston Bridge (South),53.347107,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
2,2024-12-01 00:10:00,109,20,9,True,True,True,BUCKINGHAM STREET LOWER,Buckingham Street Lower,53.353333,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
3,2024-12-01 00:10:00,11,1,29,True,True,True,EARLSFORT TERRACE,Earlsfort Terrace,53.334293,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083
4,2024-12-01 00:10:00,114,4,36,True,True,True,WILTON TERRACE (PARK),Wilton Terrace (Park),53.333652,...,0,83.2,0,0.284,0,1002.56,0,1002.26,0,0.083


In [11]:
print("Dataset Validation Checks:")
print("="*60)

# Check 1: Dataset is not empty
if df.empty:
    print("WARNING: Dataset is empty!")
else:
    print(f"Dataset contains {df.shape[0]:,} rows")

# Check 2: Check for duplicate rows
duplicate_count = df.duplicated().sum()
if duplicate_count > 0:
    print(f"WARNING: Found {duplicate_count} duplicate rows")
else:
    print("No duplicate rows found")

# Check 3: Check if all columns are unnamed
unnamed_cols = [col for col in df.columns if 'Unnamed' in str(col)]
if unnamed_cols:
    print(f"WARNING: Found {len(unnamed_cols)} unnamed columns: {unnamed_cols}")
else:
    print("All columns have names")

# Check 4: Basic column overview
print(f"\nDataset has {df.shape[1]} columns:")
print(f"  - Column names: {list(df.columns)}")



Dataset Validation Checks:
Dataset contains 298,946 rows
No duplicate rows found
All columns have names

Dataset has 78 columns:
  - Column names: ['last_reported', 'station_id', 'num_bikes_available', 'num_docks_available', 'is_installed', 'is_renting', 'is_returning', 'name', 'address', 'lat', 'lon', 'capacity', 'stno', 'year', 'month', 'day', 'hour', 'minute', 'max_air_temp_quality_indicator', 'max_air_temperature_celsius', 'min_air_temp_quality_indicator', 'min_air_temperature_celsius', 'air_temp_std_quality_indicator', 'air_temperature_std_deviation', 'max_grass_temp_quality_indicator', 'max_grass_temperature_celsius', 'min_grass_temp_quality_indicator', 'min_grass_temperature_celsius', 'grass_temp_std_quality_indicator', 'grass_temperature_std_deviation', 'max_soil_temp_5cm_quality_indicator', 'max_soil_temperature_5cm_celsius', 'min_soil_temp_5cm_quality_indicator', 'min_soil_temperature_5cm_celsius', 'soil_temp_std_5cm_quality_indicator', 'soil_temperature_std_deviation_5cm', '

Checking the validation of the row data, no duplcate rows found.

## Preparing dataset

In [12]:
import pandas as pd

# 1. Load the dataset
df = pd.read_csv('input_data/final_merged_data.csv') 
# 2. Unit Conversion: Celsius -> Kelvin
df['temp'] = df['max_air_temperature_celsius'] + 273.15

# 3. Unit Conversion: Datetime String -> Millisecond Timestamp
# Convert to datetime format first, then to int64 nanoseconds, finally to milliseconds
df['last_reported_dt'] = pd.to_datetime(df['last_reported'])
df['timestamp_ms'] = (df['last_reported_dt'].view('int64') // 10**6)

# 4. Status Conversion: True -> OPEN
df['status'] = df['is_renting'].map({True: 'OPEN', False: 'CLOSED'})

# 5. Column Renaming (Directly updating the main DataFrame columns)
rename_dict = {
    'station_id': 'number',
    'capacity': 'bike_stands',
    'lon': 'lng',
    'num_bikes_available': 'available_bikes',
    'num_docks_available': 'available_bike_stands',
    'timestamp_ms': 'last_update', # Matches time in availability data
    'last_reported': 'dt',          # Matches raw datetime string in current data
    'max_relative_humidity_percent': 'humidity',
    'max_barometric_pressure_hpa': 'pressure'
}

df_cleaned = df.rename(columns=rename_dict)

# 6. Filter final required columns 
final_columns = [
    'number', 'name', 'address', 'lat', 'lng', 'bike_stands',          # Station attributes
    'available_bikes', 'available_bike_stands', 'status', 'last_update', # Availability attributes
    'dt', 'temp', 'humidity', 'pressure'                               # Weather attributes
]

# Check if these columns exist in the DataFrame and extract them
df_final = df_cleaned[[col for col in final_columns if col in df_cleaned.columns]]

# 7. Fill in missing constant fields 
df_final['contract_name'] = 'dublin'
df_final['banking'] = 0
df_final['bonus'] = 0

# Save the processed master table
df_final.to_csv('input_data/processed_training_data.csv', index=False)

print("Data processing complete: Columns renamed, temperature converted to Kelvin, and timestamps converted to milliseconds.")

/var/folders/96/kcvffy_d1bg95g712_lbjywm0000gn/T/ipykernel_32083/3875519944.py:11: FutureWarning: Series.view is deprecated and will be removed in a future version. Use ``astype`` as an alternative to change the dtype.
  df['timestamp_ms'] = (df['last_reported_dt'].view('int64') // 10**6)
/var/folders/96/kcvffy_d1bg95g712_lbjywm0000gn/T/ipykernel_32083/3875519944.py:42: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['contract_name'] = 'dublin'
/var/folders/96/kcvffy_d1bg95g712_lbjywm0000gn/T/ipykernel_32083/3875519944.py:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pand

Data processing complete: Columns renamed, temperature converted to Kelvin, and timestamps converted to milliseconds.


Converting the feartures in final_merged_data.cvs to mathch the unit and avalibality of the accessing data (make sure the data that can be acquired later matching the traning model input feature).

In [13]:
df = pd.read_csv('input_data/processed_training_data.csv')
df.head(10)

,number,name,address,lat,lng,bike_stands,available_bikes,available_bike_stands,status,last_update,dt,temp,humidity,pressure,contract_name,banking,bonus
0,10,DAME STREET,Dame Street,53.344006,-6.266802,16,15,1,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
1,100,HEUSTON BRIDGE (SOUTH),Heuston Bridge (South),53.347107,-6.292041,25,17,8,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
2,109,BUCKINGHAM STREET LOWER,Buckingham Street Lower,53.353333,-6.249319,29,20,9,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
3,11,EARLSFORT TERRACE,Earlsfort Terrace,53.334293,-6.258503,30,1,29,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
4,114,WILTON TERRACE (PARK),Wilton Terrace (Park),53.333652,-6.248345,40,4,36,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
5,116,BROADSTONE,Broadstone,53.354700,-6.272314,30,2,28,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
6,13,FITZWILLIAM SQUARE WEST,Fitzwilliam Square West,53.336075,-6.252825,30,0,30,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
7,14,FOWNES STREET UPPER,Fownes Street Upper,53.344604,-6.263371,30,21,9,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
8,15,HARDWICKE STREET,Hardwicke Street,53.355473,-6.264423,16,1,15,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0
9,17,GOLDEN LANE,Golden Lane,53.340800,-6.267732,20,3,17,OPEN,1733011800000,2024-12-01 00:10:00,287.16,84.3,1002.56,dublin,0,0


## Training different bike availability models

In [14]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet, RidgeCV, LassoCV, ElasticNetCV


In [18]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle

# --- 1. Load Data ---
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

# Add optional weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up missing values
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# --- 4. Prepare Training Data ---
features = [
    'number',       # Station ID
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Extracted hour
    'day_of_week'   # Extracted day of the week
]

# Prediction target
target = 'available_bikes' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
model = LinearRegression()
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"LinearRegression model training completed! Prediction target: available bikes (available_bikes)")
print(f"Features used: {features}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save the model
joblib.dump(model, "process_file/bike_availability_lr_model.joblib")
with open("process_file/bike_availability_lr_model.pkl", "wb") as f:
    pickle.dump(model, f)

print("The new model has been successfully saved.")

------------------------------
LinearRegression model training completed! Prediction target: available bikes (available_bikes)
Features used: ['number', 'bike_stands', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week']
Mean Absolute Error (MAE): 7.91
R² Score: 0.0608
------------------------------
The new model has been successfully saved.


In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore') # Ignore some unnecessary warning messages

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
# Extract time features
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

# New feature: differentiate weekdays and weekends/holidays (0-4 are weekdays, 5-6 are weekends)
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features (if any)
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows with missing values in core fields
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
# Because 'number' is a category, not a numerical value. This step will turn each station into a feature column (e.g., number_42, number_85...)
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
# Define base features
base_features = [
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Hour
    'day_of_week',  # Day of the week
    'is_weekend'    # Is weekend
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

# Extract all station feature column names generated via One-Hot encoding
# pandas' get_dummies will add a prefix to the original column name, by default it's the column name itself
station_features = [col for col in df.columns if col.startswith('number_')]

# Final list of all features to be used
features = base_features + station_features

target = 'available_bikes' # Target variable: number of available bikes

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training Ridge model, please wait patiently...")
model = Ridge()
model.fit(X_train, y_train)

# --- 6. Evaluation and Saving ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Ridge model training completed! Prediction target: available bikes (available_bikes)")
print(f"Base features: {base_features}")
print(f"Number of station features: {len(station_features)}")
print(f"Total number of features: {len(features)}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save model
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_availability_ridge_model.joblib")

# Additionally save the features list, which is very important for predicting new data in the future!
# Because the feature column names changed after one-hot encoding, we must ensure the column order of the input data during prediction is exactly the same as during training
with open("process_file/bike_availability_ridge_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("New model and feature configuration file saved successfully.")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training Ridge model, please wait patiently...
Evaluating model...
------------------------------
Ridge model training completed! Prediction target: available bikes (available_bikes)
Base features: ['bike_stands', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week', 'is_weekend']
Number of station features: 115
Total number of features: 122
Mean Absolute Error (MAE): 6.26
R² Score: 0.3473
------------------------------
Saving model and feature list...
New model and feature configuration file saved successfully.


In [20]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore') 

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
# Extract time features
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

# New feature: distinguish between weekdays and weekends/holidays (0-4 are weekdays, 5-6 are weekends)
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features (if any)
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows with missing values in core fields
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# One-hot encode the station numbers
# Because 'number' is a categorical variable, not a numerical magnitude. This step will turn each station into a feature column (e.g., number_42, number_85...)
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
# Define base features (removed the original 'number')
base_features = [
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Hour
    'day_of_week',  # Day of the week
    'is_weekend'    # Is weekend
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

# Extract all station feature column names generated by One-Hot encoding
# pandas' get_dummies will add a prefix to the original column name, defaulting to the column name itself
station_features = [col for col in df.columns if col.startswith('number_')]

# Final list of all features to be used
features = base_features + station_features

target = 'available_bikes' # Target variable: number of available bikes

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
# Note: The original print statement says "Random Forest" but the code uses RidgeCV. Translated exactly as written.
print("Training RidgeCV model, please be patient (may take a few minutes)...")
model = RidgeCV()
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"RidgeCV model training complete! Prediction target: number of available bikes (available_bikes)")
print(f"Base features: {base_features}")
print(f"Number of station features: {len(station_features)}")
print(f"Total number of features: {len(features)}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save the model
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_availability_ridgeCV_model.joblib")

# Save the features list separately; this is very important for predicting new data in the future!
# Because the feature column names change after one-hot encoding, we must ensure the column order of the input data during prediction is exactly the same as during training
with open("process_file/bike_availability_ridgeCV_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("New model and feature configuration files saved successfully.")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training RidgeCV model, please be patient (may take a few minutes)...
Evaluating model...
------------------------------
RidgeCV model training complete! Prediction target: number of available bikes (available_bikes)
Base features: ['bike_stands', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week', 'is_weekend']
Number of station features: 115
Total number of features: 122
Mean Absolute Error (MAE): 6.26
R² Score: 0.3473
------------------------------
Saving model and feature list...
New model and feature configuration files saved successfully.


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor 
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore') 

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
# Extract time features
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

# New feature: distinguish between weekdays and weekends/holidays (0-4 are weekdays, 5-6 are weekends)
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features (if any)
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows with missing values in core fields
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
# Because 'number' is a category, not a numerical value. This step will turn each station into a feature column (e.g., number_42, number_85...)
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
# Define base features
base_features = [
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Hour
    'day_of_week',  # Day of the week
    'is_weekend'    # Is weekend
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

# Extract all station feature column names generated by One-Hot Encoding
# pandas' get_dummies will add a prefix to the original column names, which defaults to the column name itself
station_features = [col for col in df.columns if col.startswith('number_')]

# Final list of all features to be used
features = base_features + station_features

target = 'available_bikes' # Target variable: number of available bikes

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training the Random Forest model, please wait patiently (may take a few minutes)...")
model = RandomForestRegressor(n_estimators=100, max_depth=15, n_jobs=-1, random_state=42)
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating the model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Random Forest model training completed! Prediction target: available bikes (available_bikes)")
print(f"Base features: {base_features}")
print(f"Number of station features: {len(station_features)}")
print(f"Total number of features: {len(features)}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save the model
print("Saving the model and feature list...")
joblib.dump(model, "process_file/bike_availability_rf_model.joblib")

# Additionally save the features list, which is very important when predicting new data in the future!
# Because the feature column names change after one-hot encoding, we must ensure the column order of input data during prediction is exactly the same as during training
with open("process_file/bike_availability_rf_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("The new model and feature configuration files have been saved successfully.")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training the Random Forest model, please wait patiently (may take a few minutes)...
Evaluating the model...
------------------------------
Random Forest model training completed! Prediction target: available bikes (available_bikes)
Base features: ['bike_stands', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week', 'is_weekend']
Number of station features: 115
Total number of features: 122
Mean Absolute Error (MAE): 6.14
R² Score: 0.3323
------------------------------
Saving the model and feature list...
The new model and feature configuration files have been saved successfully.


In [23]:
import pandas as pd
import time
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor       
from sklearn.preprocessing import StandardScaler      
from sklearn.pipeline import Pipeline                
from sklearn.metrics import mean_absolute_error, r2_score
import joblib


# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
# Extract time features
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

# Add new feature: distinguish between weekdays and weekends/holidays (0-4 are weekdays, 5-6 are weekends)
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features (if any)
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows with missing values in core fields
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
# Because 'number' is a categorical variable, not a numerical magnitude. This step turns each station into a feature column (e.g., number_42, number_85...)
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
# Define base features
base_features = [
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Hour
    'day_of_week',  # Day of the week
    'is_weekend'    # Is weekend
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

# Extract all station feature column names generated by One-Hot encoding
# pandas' get_dummies will add a prefix to the original column name, which defaults to the column name itself
station_features = [col for col in df.columns if col.startswith('number_')]

# Final list of all features used
features = base_features + station_features

target = 'available_bikes' # Target variable: number of available bikes

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)






print("-" * 30)
print("Building Neural Network (MLP) pipeline...")

# [Core Configuration]: Set up the pipeline, standardize first, then feed into the neural network
model_pipeline = Pipeline([
    ('scaler', StandardScaler()),  
    # Configure MLP model parameters
    ('mlp', MLPRegressor(
        hidden_layer_sizes=(128, 64, 32), # Hidden layer architecture: 3 layers, number of neurons decreases like a "funnel"
        activation='relu',                # Activation function: ReLU is the go-to for deep learning
        solver='adam',                    # Optimization algorithm: Adam usually converges the fastest
        alpha=0.001,                      # L2 regularization penalty coefficient to prevent overfitting
        batch_size=256,                   # Batch size: take 256 samples each time to update weights
        max_iter=500,                     # Maximum number of iterations (training epochs)
        early_stopping=True,              # [Crucial] Early stopping mechanism: if the model finds it can't learn anymore, it stops early to save time!
        random_state=42
    ))
])

print("Training the Neural Network model, please wait patiently...")
start_time = time.time()

# Start training
model_pipeline.fit(X_train, y_train)

end_time = time.time()
print(f"Training completed! Time taken: {end_time - start_time:.1f} seconds")

# --- Evaluate and Save ---
print("Evaluating the model...")
y_pred = model_pipeline.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"MLP Neural Network Evaluation Results:")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print(f"Actual iterations: {model_pipeline.named_steps['mlp'].n_iter_}") # See how many epochs it ran before early stopping
print("-" * 30)

# Save the entire pipeline
joblib.dump(model_pipeline, "output_model/bike_availability_mlp_pipeline.joblib")
print("MLP pipeline saved successfully!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Building Neural Network (MLP) pipeline...
Training the Neural Network model, please wait patiently...
Training completed! Time taken: 104.8 seconds
Evaluating the model...
------------------------------
MLP Neural Network Evaluation Results:
Mean Absolute Error (MAE): 1.94
R² Score: 0.9189
Actual iterations: 138
------------------------------
MLP pipeline saved successfully!


In [24]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore') 

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
# Extract time features
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# ================= Construct Lag Features =================
# Sort strictly by station ID and time first to prevent data mixing!
df = df.sort_values(by=['number', 'timestamp'])

# Group by and shift down one row to get the bike count of the "previous record" for that station
df['bikes_1h_ago'] = df.groupby('number')['available_bikes'].shift(1)

# Drop the null values in the first row generated by shift(1) (the first record of each station has no historical data)
df.dropna(subset=['bikes_1h_ago'], inplace=True)
# ==========================================================

# Handle missing weather features (if any)
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows with missing values in core fields
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bikes', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
# This step will turn each station into a feature column (e.g., number_42, number_85...)
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
# Define base features
base_features = [
    'bike_stands',  # Total station capacity
    'bikes_1h_ago', # Number of bikes in the previous hour
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Hour
    'day_of_week',  # Day of the week
    'is_weekend'    # Is weekend
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

# Extract all station feature column names generated by One-Hot Encoding
station_features = [col for col in df.columns if col.startswith('number_')]

# Final list of all features used (base features + dummy variables for all stations)
features = base_features + station_features

target = 'available_bikes' # Target variable: available bikes

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training the Random Forest model, please wait patiently...")

model = RandomForestRegressor(n_estimators=100, max_depth=15, n_jobs=-1, random_state=42)
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating the model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Training of Random Forest model with lag features complete! Prediction target: {target}")
print(f"Base features: {base_features}")
print(f"Number of station features: {len(station_features)}")
print(f"Total number of features: {len(features)}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save the model 
print("Saving the model and feature list...")
joblib.dump(model, "process_file/bike_availability_rf2_model_final.joblib")

# The feature list must be saved! When predicting with new data in the future, it must be fed to the model in the exact order of this list
with open("process_file/bike_availability_rf2_model_final.pkl", "wb") as f:
    pickle.dump(features, f)

print("Model and feature configuration files have been successfully saved!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training the Random Forest model, please wait patiently...
Evaluating the model...
------------------------------
Training of Random Forest model with lag features complete! Prediction target: available_bikes
Base features: ['bike_stands', 'bikes_1h_ago', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week', 'is_weekend']
Number of station features: 115
Total number of features: 123
Mean Absolute Error (MAE): 0.61
R² Score: 0.9815
------------------------------
Saving the model and feature list...
Model and feature configuration files have been successfully saved!


## Comparing bike availability model results

Several base features are picked based on the data access availability and the meaningness:'bike_stands', 'temp', 'humidity', 'pressure', 'hour'(engineered), 'day_of_week'(engineered),'is_weekend'(engineered). 

And the feature 'number'(station number) is a categrecal feature, using one-hot encoding method to transform the feature.

The target feature is 'available_bikes'(numerical).

In the last RandomForestRegressor model, the feature 'bikes_1h_ago' is added, but the R square is too high (0.9815) which means the model is overfitting, so this model was discarded.

Several models are use to make to train the model: LinearRegression, RandomForestRegressor, Ridge, RidgeCV, MLP. 

The model that has reasonable best performance is MLP, which R suqare is 0.9189 and the MAE is 1.94.

## Training different bike stands availability models

In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle

# --- 1. Load Data ---
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek

if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up missing values
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# --- 4. Prepare Training Data ---
features = [
    'number',       # Station ID
    'bike_stands',  # Total station capacity
    'temp',         # Temperature
    'humidity',     # Humidity
    'pressure',     # Pressure
    'hour',         # Extracted hour
    'day_of_week'   # Extracted day of the week
]

if 'wind_speed' in df.columns:
    features.append('wind_speed')
if 'rain_1h' in df.columns:
    features.append('rain_1h')

# Prediction target is the number of available bike stands
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
model = LinearRegression()
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"LinearRegression available stands prediction model training completed!")
print(f"Features used: {features}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Export model files
joblib.dump(model, "process_file/bike_stands_availability_lr_model.joblib")
with open("process_file/bike_stands_availability_lr_model.pkl", "wb") as f:
    pickle.dump(model, f)

print("New model successfully saved.")

------------------------------
LinearRegression available stands prediction model training completed!
Features used: ['number', 'bike_stands', 'temp', 'humidity', 'pressure', 'hour', 'day_of_week']
Mean Absolute Error (MAE): 8.05
R² Score: 0.2300
------------------------------
New model successfully saved.


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore')

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Datetime Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean invalid rows (drop missing values)
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
base_features = [
    'bike_stands',   # Total station capacity
    'temp',          
    'humidity',      
    'pressure',      
    'hour',          
    'day_of_week',   
    'is_weekend'     
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

station_features = [col for col in df.columns if col.startswith('number_')]
features = base_features + station_features

# Target variable: available bike stands
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training Ridge model, please wait patiently...")
model = Ridge()
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Ridge model training complete! Prediction target: {target}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save the model (added _no_lag suffix for distinction)
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_stands_ridge_model.joblib")

with open("process_file/bike_stands_ridge_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("Model and feature configuration files successfully saved!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training Ridge model, please wait patiently...
Evaluating model...
------------------------------
Ridge model training complete! Prediction target: available_bike_stands
Mean Absolute Error (MAE): 6.27
R² Score: 0.4855
------------------------------
Saving model and feature list...
Model and feature configuration files successfully saved!


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore')

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Datetime Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Drop invalid rows 
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# One-Hot Encode station numbers
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
base_features = [
    'bike_stands',   # Total station capacity
    'temp',          
    'humidity',      
    'pressure',      
    'hour',          
    'day_of_week',   
    'is_weekend'     
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

station_features = [col for col in df.columns if col.startswith('number_')]
features = base_features + station_features

# Target variable: Available bike stands
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training RidgeCV model, please wait...")
model = RidgeCV()
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"RidgeCV model training completed! Prediction target: {target}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save model 
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_stands_ridgeCV_model.joblib")

with open("process_file/bike_stands_ridgeCV_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("RidgeCV model and feature configuration file saved successfully!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training RidgeCV model, please wait...
Evaluating model...
------------------------------
RidgeCV model training completed! Prediction target: available_bike_stands
Mean Absolute Error (MAE): 6.27
R² Score: 0.4855
------------------------------
Saving model and feature list...
RidgeCV model and feature configuration file saved successfully!


In [29]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore')

# --- 1. Load data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature engineering ---
print("Performing feature engineering...")
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean invalid rows
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# One-Hot Encoding for station numbers
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare training data ---
print("Preparing model input data...")
base_features = [
    'bike_stands',   # Total station capacity
    'temp',          
    'humidity',      
    'pressure',      
    'hour',          
    'day_of_week',   
    'is_weekend'     
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

station_features = [col for col in df.columns if col.startswith('number_')]
features = base_features + station_features

# Target variable: available bike stands
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train model ---
print("-" * 30)
print("Training Random Forest model, please wait patiently...")
model = RandomForestRegressor(n_estimators=100, max_depth=15, n_jobs=-1, random_state=42)
model.fit(X_train, y_train)

# --- 6. Evaluate and save ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Random Forest model training complete! Prediction target: {target}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save model
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_stands_rf_model.joblib")

with open("process_file/bike_stands_rf_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("Random Forest model and feature configuration file saved successfully!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training Random Forest model, please wait patiently...
Evaluating model...
------------------------------
Random Forest model training complete! Prediction target: available_bike_stands
Mean Absolute Error (MAE): 5.63
R² Score: 0.5427
------------------------------
Saving model and feature list...
Random Forest model and feature configuration file saved successfully!


In [30]:
import pandas as pd
import time
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor       
from sklearn.preprocessing import StandardScaler      
from sklearn.pipeline import Pipeline                 
from sklearn.metrics import mean_absolute_error, r2_score
import joblib


# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Handle missing weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean invalid rows 
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
base_features = [
    'bike_stands',   # Total station capacity
    'temp',          
    'humidity',      
    'pressure',      
    'hour',          
    'day_of_week',   
    'is_weekend'     
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

station_features = [col for col in df.columns if col.startswith('number_')]
features = base_features + station_features

# Target variable: Number of available bike stands 
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


print("-" * 30)
print("Building Neural Network (MLP) pipeline...")

# Set up pipeline: standardize first, then pass to neural network
model_pipeline = Pipeline([
    ('scaler', StandardScaler()),  
    # Configure MLP model parameters
    ('mlp', MLPRegressor(
        hidden_layer_sizes=(128, 64, 32), # Hidden layer architecture: 3 layers, neuron count decreases in a "funnel" shape
        activation='relu',                # Activation function: ReLU is the Swiss Army knife of deep learning
        solver='adam',                    # Optimization algorithm: Adam usually converges the fastest
        alpha=0.001,                      # L2 regularization penalty term to prevent overfitting
        batch_size=256,                   # Batch size: update weights using 256 samples each time
        max_iter=500,                     # Maximum number of iteration (training) rounds
        early_stopping=True,              # [Key] Early stopping mechanism: stops early to save time if the model stops improving!
        random_state=42
    ))
])

print("Training the Neural Network model, please wait patiently...")
start_time = time.time()

# Start training
model_pipeline.fit(X_train, y_train)

end_time = time.time()
print(f"Training completed! Time elapsed: {end_time - start_time:.1f} seconds")

# --- Evaluation and Saving ---
print("Evaluating the model...")
y_pred = model_pipeline.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"MLP Neural Network Evaluation Results:")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print(f"Actual iterations: {model_pipeline.named_steps['mlp'].n_iter_}") # See how many rounds it ran before early stopping
print("-" * 30)

# Save the entire pipeline
joblib.dump(model_pipeline, "output_model/bike_stands_mlp_pipeline.joblib")
print("MLP pipeline saved successfully!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Building Neural Network (MLP) pipeline...
Training the Neural Network model, please wait patiently...
Training completed! Time elapsed: 93.0 seconds
Evaluating the model...
------------------------------
MLP Neural Network Evaluation Results:
Mean Absolute Error (MAE): 2.01
R² Score: 0.9327
Actual iterations: 127
------------------------------
MLP pipeline saved successfully!


In [31]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score
import joblib
import pickle
import warnings
warnings.filterwarnings('ignore')

# --- 1. Load Data ---
print("Loading data...")
df = pd.read_csv("input_data/processed_training_data.csv")

# --- 2. Time Processing ---
df['timestamp'] = pd.to_datetime(df['dt'])

# --- 3. Feature Engineering ---
print("Performing feature engineering...")
df['hour'] = df['timestamp'].dt.hour
df['day_of_week'] = df['timestamp'].dt.dayofweek
df['is_weekend'] = df['day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# ================= Build Lag Features (for available stands) =================
df = df.sort_values(by=['number', 'timestamp'])

# Number of available stands 1 hour ago
df['stands_1h_ago'] = df.groupby('number')['available_bike_stands'].shift(1)

# Drop the first row of null values generated by shift(1)
df.dropna(subset=['stands_1h_ago'], inplace=True)
# ==========================================================

# Handle missing weather features
if 'rain_1h' in df.columns:
    df['rain_1h'] = df['rain_1h'].fillna(0)
if 'snow_1h' in df.columns:
    df['snow_1h'] = df['snow_1h'].fillna(0)

# Clean up invalid rows
df.dropna(subset=['temp', 'humidity', 'pressure', 'available_bike_stands', 'bike_stands'], inplace=True)

# Perform One-Hot Encoding on station numbers
df = pd.get_dummies(df, columns=['number'], drop_first=False)

# --- 4. Prepare Training Data ---
print("Preparing model input data...")
base_features = [
    'bike_stands',   # Total station capacity 
    'stands_1h_ago', # Use the number of available stands from 1 hour ago as a feature
    'temp',          
    'humidity',      
    'pressure',      
    'hour',          
    'day_of_week',   
    'is_weekend'     
]

# Add optional weather features
if 'wind_speed' in df.columns:
    base_features.append('wind_speed')
if 'rain_1h' in df.columns:
    base_features.append('rain_1h')

station_features = [col for col in df.columns if col.startswith('number_')]
features = base_features + station_features

# Target variable
target = 'available_bike_stands' 

X = df[features]
y = df[target]

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# --- 5. Train Model ---
print("-" * 30)
print("Training available stands prediction model with lag features (Random Forest), please wait patiently...")
model = RandomForestRegressor(n_estimators=100, max_depth=15, n_jobs=-1, random_state=42)
model.fit(X_train, y_train)

# --- 6. Evaluate and Save ---
print("Evaluating model...")
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("-" * 30)
print(f"Available stands prediction model training completed! Target: {target}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R² Score: {r2:.4f}")
print("-" * 30)

# Save files
print("Saving model and feature list...")
joblib.dump(model, "process_file/bike_stands_rf2_model.joblib")

with open("process_file/bike_stands_rf2_model.pkl", "wb") as f:
    pickle.dump(features, f)

print("Available stands prediction model with lag features and feature configuration file saved successfully!")

Loading data...
Performing feature engineering...
Preparing model input data...
------------------------------
Training available stands prediction model with lag features (Random Forest), please wait patiently...
Evaluating model...
------------------------------
Available stands prediction model training completed! Target: available_bike_stands
Mean Absolute Error (MAE): 0.62
R² Score: 0.9854
------------------------------
Saving model and feature list...
Available stands prediction model with lag features and feature configuration file saved successfully!


## Comparing bike stands availability model results

Several base features are picked based on the data access availability and the meaningness:'bike_stands', 'temp', 'humidity', 'pressure', 'hour'(engineered), 'day_of_week'(engineered),'is_weekend'(engineered). 

And the feature 'number'(station number) is a categrecal feature, using one-hot encoding method to transform the feature.

The target feature is 'available_bike_stands'(numerical).

In the last RandomForestRegressor model, the feature 'stands_1h_ago' is added, but the R square is too high (0.9854) which means the model is overfitting, so this model was discarded.

Several models are use to make to train the model: LinearRegression, RandomForestRegressor, Ridge, RidgeCV, MLP. 

The model that has reasonable best performance is MLP, which R suqare is 0.9327 and the MAE is 2.01.